### PR TITLE
fix(obstacle_avoidance_planner): solution of changing the sampling distance of behavior velocity planner causing obstacle avoidance planner to die

### PR DESCRIPTION
**Pull request by shulanbushangshu**
## Description
When changing the sampling length in the"behavior_velocity_planner" node, the "obstacle_avoidance_planner"node will die.
The length in the "obstacle_avoidance_planner" node is 5 (fixed). When it is changed to a value greater than 5 in the "behavior_velocity_planner"node, the "obstacle_avoidance_planner" node will die when the vehicle reaches the destination.

## Related links
https://github.com/autowarefoundation/autoware.universe/issues/798

## Tests performed
Current simulation:
obs_die

Modified simulation:
obs_nodie

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by shulanbushangshu** [comment]
@kosuke55

File: planning/obstacle_avoidance_planner/src/node.cpp [hunk 1]
```diff
@@ -1499,6 +1499,7 @@ ObstacleAvoidancePlanner::alignVelocity(
   size_t prev_begin_idx = 0;
   for (size_t i = 0; i < fine_traj_points_with_vel.size(); ++i) {
     const auto truncated_points = points_utils::clipForwardPoints(path_points, prev_begin_idx, 5.0);
+    if (truncated_points.size() < 1) continue;
```
Comment (taikitanaka3) [inline planning/obstacle_avoidance_planner/src/node.cpp:1502]:
```suggestion
    if (truncated_points.empty()) continue;
```

**Comment by taikitanaka3** [comment]
@shulanbushangshu 
Thank you for your PR. Your branch looks this PR is not included https://github.com/autowarefoundation/autoware.universe/pull/1320
can you include [this PR](https://github.com/autowarefoundation/autoware.universe/pull/1320) and test this again?

**Comment by shulanbushangshu** [comment]
@taikitanaka3 .Thank you for your review,I  will confirm the code and retest

**Comment by shulanbushangshu** [comment]
@taikitanaka3 ,I have confirmed the code and retest:
velocity_change
die_ao

Modified simulation:

nodie

**Comment by taikitanaka3** [comment]
@shulanbushangshu 
Thank you. Modified simulation will not die at obstacle avoidance planner but dies at behavior velocity planner?

File: planning/obstacle_avoidance_planner/src/node.cpp (hunk 1 above)
Comment (taikitanaka3) [reply planning/obstacle_avoidance_planner/src/node.cpp:1499]: @shulanbushangshu 
Thank you for testing. But I  re-think my [fix PR](https://github.com/autowarefoundation/autoware.universe/pull/1369#issuecomment-1188763184) is irrelevant to https://github.com/autowarefoundation/autoware.universe/issues/798 so can you keep confirm this PR fix the issue?

**Comment by taikitanaka3** [comment]
@shulanbushangshu 
Could you please re-test including my fix PR 
with / without your fix PR with sampling interval greater than 5.0 ?

[Is this comment](https://github.com/autowarefoundation/autoware.universe/pull/1369#issuecomment-1188814984) doing above?

**Comment by taikitanaka3** [comment]
@shulanbushangshu I've tested with this PR as you suggests ego can't goal
without this PR
image
with this PR
image

Do you have any good suggestion to arrive goal?

**Comment by takayuki5168** [comment]
Thank you for your discussion.
I'll look into the problem to realize both reaching the goal and the node keeping alive. Please wait a bit.

**Comment by shulanbushangshu** [comment]
@taikitanaka3 @takayuki5168 
Thank you for your review.I fix the problem of not reaching the goal.Can you check it?
obs_avoid_ok

**Comment by taikitanaka3** [comment]
@shulanbushangshu 
Thank you , but build is failed because of writing const value.
@takayuki5168 
FYI  with this change I confirm this problem is fixed
image

File: planning/obstacle_avoidance_planner/src/node.cpp [hunk 2]
```diff
@@ -1486,6 +1486,9 @@ ObstacleAvoidancePlanner::alignVelocity(
   size_t prev_begin_idx = 0;
   for (size_t i = 0; i < fine_traj_points_with_vel.size(); ++i) {
     const auto truncated_points = points_utils::clipForwardPoints(path_points, prev_begin_idx, 5.0);
```
Comment (taikitanaka3) [inline planning/obstacle_avoidance_planner/src/node.cpp:1488]:
```suggestion
    auto truncated_points = points_utils::clipForwardPoints(path_points, prev_begin_idx, 5.0);
```

Comment (taikitanaka3) [reply planning/obstacle_avoidance_planner/src/node.cpp:1488]: I think `truncated_points.assign(path_points.begin(), path_points.end());` is O
but I found an odd behavior please wait for murooka-san
https://user-images.githubusercontent.com/65527974/179894007-63cfcd49-b4a1-47b3-8e65-1be2bfa0add3.mp4

**Comment by shulanbushangshu** [comment]
Yes,I forget to change the "const".I fix it.

**Comment by taikitanaka3** [comment]
@shulanbushangshu 
If i am not wrong this bug is only reproduces with setting behavior velocity planner sampling interval more than  5m,
and that setting is not suitable for autonomous driving. If this problem happens for normal driving case with sampling interval 1.0m this bug is high priority, if not I think we can set this PR low priority. what do you think?

**Comment by shulanbushangshu** [comment]
@taikitanaka3 OK,we can set this PR low priority

**Comment by takayuki5168** [comment]
@shulanbushangshu @xmfcx @Sharrrrk cc @mitsudome-r 
Sorry to be late. Your change can solve the issue, but I guess there may be a better code. Let me implement it, and I'll let you know in a few days.

**Comment by takayuki5168** [comment]
@shulanbushangshu @taikitanaka3 
I created another PR to make commit history clear. Could you review this.
https://github.com/autowarefoundation/autoware.universe/pull/1616

**Comment by takayuki5168** [comment]
@shulanbushangshu @xmfcx @Sharrrrk 
We can close this PR since I merged another PR for the issue. Let me know if there is another bug.
https://github.com/autowarefoundation/autoware.universe/pull/1616

**Comment by shulanbushangshu** [comment]
@takayuki5168 
Yes,I have simulated the new code and it works well.I will close this PR.

**Pull request closed without merging**